pkg/urbit/include/vere/epoc.h

```diff
@@ -0,0 +1,272 @@
+u3_epoc*
+u3_epoc_new(const c3_path* const par_u, const c3_d fir_d);
+
+/// Create a new first epoch from an existing (non-epoch-based) event log.
+///
+/// @note This is intended for use in migration from the non-epoch-based event
+///       log to the new epoch-based event log.
+///
+/// @param[in]  src_u  Path to directory housing existing event log.
+/// @param[in]  par_u  Parent directory to house epoch. Will be created if it
+///                    doesn't already exist.
+/// @param[in]  cur_d  Event number of most recently applied event.
+///
+/// @return NULL  New epoch could not be created from old event log.
+/// @return       Handle to migrated epoch.
+u3_epoc*
+u3_epoc_migrate(c3_path* const       src_u,
+                const c3_path* const par_u,
+                const c3_d           cur_d);
+
+/// Load an existing epoch created with u3_epoc_new().
+///
+/// @param[in]  pax_u  Root directory of epoch.
+/// @param[in]  rdo_t  If non-zero, the epoch will be inert (a strict form of
+///                    read-only), meaning that only the epoch's first and last
+///                    event numbers and path will be accessible. In particular,
+///                    u3_epoc_commit() and u3_epoc_iter_*() must NOT be called.
+/// @param[out] len_w  Pointer to length of boot sequence. Only relevant for
+///                    first epoch. Can be NULL if not first epoch. If not NULL
+///                    and not first epoch, will be set to 0.
+///
+/// @return NULL  Existing epoch could not be opened.
+/// @return       Handle to open epoch.
+u3_epoc*
+u3_epoc_open(const c3_path* const pax_u, const c3_t rdo_t, c3_w* const len_w);
+
+/// Get the file path of an epoch.
+///
+/// @param[in] poc_u  Epoch handle. Must not be NULL.
+///
+/// @return  File path.
+const c3_path*
+u3_epoc_path(const u3_epoc* const poc_u);
+
+/// Get the string representation of an epoch's file path.
+///
+/// @param[in] poc_u  Epoch handle. Must not be NULL.
+///
+/// @return  String representation of file path.
+const c3_c*
+u3_epoc_path_str(const u3_epoc* const poc_u);
+
+/// Get the ID of the first committed event in an epoch.
+///
+/// @param[in] poc_u  Epoch handle. Must have at least one committed event and
+///                   must not be NULL.
+///
+/// @return  ID of epoch's first committed event.
+c3_d
+u3_epoc_first_commit(const u3_epoc* const poc_u);
+
+/// Get the ID of the last committed event in an epoch.
+///
+/// @param[in] poc_u  Epoch handle. Must have at least one committed event and
+///                   must not be NULL.
+///
+/// @return  ID of epoch's last committed event.
+c3_d
+u3_epoc_last_commit(const u3_epoc* const poc_u);
+
+/// Get the ID of the first event represented by the epoch, which is equivalent
+/// to the ID of the event that corresponds to the epoch's snapshot.
+///
+/// @param[in] poc_u  Epoch handle. Must not be NULL.
+///
+/// @return  ID of epoch's first event.
+c3_d
+u3_epoc_first_evt(const u3_epoc* const poc_u);
```

Comment:
This API is a little confusing. "first_evt" is actually just the epoch number, and it refers to the initial state of the epoch (stored in the snapshot, caused by the last event of the preceding epoch). First/last "commit" is actually the first and last event in the epoch.